File: java/src/main/java/com/showmeyourcode/playground/java/overview/datatype/Datatypes.java
package com.showmeyourcode.playground.java.overview.datatype;

import com.showmeyourcode.playground.kotlin.common.Descriptions;
import lombok.extern.slf4j.Slf4j;


@Slf4j
public class Datatypes {

    private Datatypes(){}

    public static void main(String[] args) {
        log.info("{}", Descriptions.header(Descriptions.DATA_TYPE_TITLE));
        log.info(Descriptions.DATA_TYPE);

        objectType();
        signedUnsignedVariables();
        typeInference();
        rawType();
    }

    private static void rawType() {
        log.info("\n{} Raw Type", Descriptions.INDENT1);
        log.info("If the actual type argument is omitted, you create a raw type of Box<T>: Box rawBox = new Box();");
        log.info("A list is a raw type, while List<String> is a parameterized type.");
        log.info("""
                
                Note that the reason why raw types exist is for backwards compatibility with Java 1.4 and older,
                which did not have generics at all.
                """
        );
        // https://docs.oracle.com/javase/tutorial/java/generics/rawTypes.html
        // https://stackoverflow.com/questions/2770321/what-is-a-raw-type-and-why-shouldnt-we-use-it
        log.info("""
                Using raw types is still possible, they should be avoided:           
                - They usually require casts.
                - They aren't type safe, and some important kinds of errors will only appear at runtime.
                - They are less expressive, and don't self-document in the same way as parameterized types..
                """
        );
    }

    private static void typeInference() {
        log.info("\n{} Type Inference", Descriptions.INDENT1);
        log.info("Java does have type inferencing when using generics.");
        log.info("""
                With var , the Java compiler infers the type of the variable at compile time,
                using type information obtained from the variable's initializer.
                """
        );
    }

    private static void objectType() {
        log.info("Primitive data types are predefined. Object data types are user-defined.");
        log.info("Primitives simply represent a value, like the number seven or the boolean value of false.");
        log.info("They have no methods and no complementary properties.");
        log.info("Also, you can't do much with them other than inspect their value or perhaps change the value that they hold.");
        log.info("But they are not, by any means, objects.\n");
        log.info("Object == 1.2L:       {}", false);
        log.info("Object == \"String\":   {}", "String" instanceof Object);
        log.info("Object == 'a':        {}", false);
        log.info("Object == 'int[]':    {}", new int[]{1, 2, 3} instanceof Object);
        log.info("Object == 'Integer':  {}", Integer.valueOf(1) instanceof Object);
        log.info("Object == 'Integer':  {}", Integer.valueOf(1) instanceof Object);
    }

    private static void signedUnsignedVariables() {
        // https://stackoverflow.com/questions/21089540/signed-and-unsigned-data-types-in-java
        log.info("\n{} Signed/Unsigned variables", Descriptions.INDENT1);
        log.info("Signed variables, such as signed integers will allow you to represent numbers both in the positive and negative ranges.");
        log.info("Unsigned variables, such as unsigned integers, will only allow you to represent numbers in the positive.\n");

        log.info("Java only supports signed types (except char), in other words all Java numeric types are signed. This was the designers decision.");
        log.info("'0' * 1.1: {}", '0' * 1.1);

        // https://stackoverflow.com/questions/430346/why-doesnt-java-support-unsigned-ints/15053268#15053268
        log.info("""
            Update! With JDK8 there are new APIs for Long and Integer which provide helper methods
            when treating long and int values as unsigned values.
            """
        );

        int w = 150;
        int z = 10;

        log.info("Integer.divideUnsigned(150, 10): {}", Integer.divideUnsigned(w, z));

        int m = -150;
        int n = 10;

        log.info("Integer.divideUnsigned(-150, 10): {}", Integer.divideUnsigned(m, n));
    }
}

File: kotlin/src/main/kotlin/com/showmeyourcode/playground/kotlin/overview/datatype/Datatypes.kt
package com.showmeyourcode.playground.kotlin.overview.datatype

import com.showmeyourcode.playground.kotlin.common.Descriptions
import com.showmeyourcode.playground.kotlin.common.Logging

object Datatypes {
    fun main() {
        Logging.LOGGER.info("\n${Descriptions.DATA_TYPE_HEADER}")
        Logging.LOGGER.info(Descriptions.DATA_TYPE)

        // https://kotlinlang.org/docs/basic-types.html
        Logging.LOGGER.info(
            """
            |In Kotlin, everything is an object in the sense 
            |that you can call member functions and properties on any variable.
            """.trimIndent()
        )

        types()
        primitives()
        signedUnsignedVariables()
    }

    private fun primitives() {
        Logging.LOGGER.info("\nDoes Kotlin have primitive types? No, it does not.")
        Logging.LOGGER.info("Primitives are a silly construct, which Kotlin does not attempt to bring in from Java.")
        Logging.LOGGER.info(
            """Everything in Kotlin is a class and Int is no exception, 
                |leading to a completely unified type system where everything inherits from Any.
            """.trimMargin()
        )
        // https://kotlinlang.org/docs/comparison-to-java.html#what-java-has-that-kotlin-does-not
        // https://discuss.kotlinlang.org/t/does-kotlin-have-primitive-types/13789/4
    }

    private fun types() {
        // https://mfallahpour.medium.com/any-nothing-and-unit-in-kotlin-6b0b9779c197
        Logging.LOGGER.info(
            "Any is the root of the Kotlin class hierarchy. Every Kotlin class has Any as a superclass.\n"
        )

        Logging.LOGGER.info(
            """Nothing, you can use Nothing to represent 'a value that never exists': 
               |for example, if a function has the return type of Nothing, it means that it never returns 
               |(always throws an exception).
            """.trimMargin()
        )

        Logging.LOGGER.info(
            """
                |
                |Unit is the same thing as void in Java or C. 
                |When the return type of a function is Unit it means that the function does not return anything."""
                .trimMargin()
        )

        Logging.LOGGER.info(
            """
                |
                |Local type inference in Kotlin is the process of deducing the compile-time types of expressions, 
                |lambda expression parameters and properties. Ref: https://kotlinlang.org/spec/type-inference.html
            """.trimMargin()
        )

        val typeInterferenceExample = "1".toIntOrNull() ?: 0
        Logging.LOGGER.info(
            """
                |val a = \"1\".toIntOrNull() ?: 0: $typeInterferenceExample 
                |-> you don't have to specify the type as it's deducted by Kotlin/Type inference
            """.trimMargin()
        )
    }

    private fun signedUnsignedVariables() {
        // https://www.baeldung.com/kotlin/unsigned-integers
        Logging.LOGGER.info("\n====> Signed/Unsigned variables")
        Logging.LOGGER.info("As of Kotlin 1.3, Kotlin supports unsigned integers to accommodate this requirement.")

        val uByte: UByte = 42u
        val uShort: UShort = 42u
        val uInt: UInt = 42U
        val uLong: ULong = 42u

        Logging.LOGGER.info("UByte: $uByte UShort: $uShort UInt: $uInt ULong: $uLong (42u * 43u): ${42u * 43u}")
    }
}
